Profiler: Better structures, count calls, measure relative usage

t4im opened the pull request:
Follow-up of #4245

#### Changes
The first commit refactors the code for better extensibility and improved performance.
- Allow to log more types of measurement without having to add new lookup tables.
- Improve sampling and measurement performance.
- Have Stats initialize automatically.
- Make labeling and its reporting more flexible.
- Isolate data storing/handling from sampling.
- Some aesthetically changes, like dynamic first column width
- pass along return values from callback registration functions for increased future-safety

The second commit changes the measurements taken to avoid misleading results from profiling sessions.
- Count how often a function is called and report an average call per active sample in txt (total in csv)
- Replace "relative time spend per server step" with "relative usage".
  > Because we already measure absolute time, the relative time did not provide additional useful information. "relative usage" shows how often an instrument was sampled relative to all samples (csv provides absolute numbers). That way you know if an instrument is often idle or frequently used, which makes a difference for other measures.
- Don't discard 0-time measurements anymore, only negative measurements
  > Instrumentalized functions that return early should not unnecessarily be penalized as being more heavyweight just because we skip their 0-time measurments.

#### Example console dump

| instrumentation | calls | use % | avg µs | min µs | max µs |
| --- | --- | --- | --- | --- | --- |
| pipeworks: | 8.2 | 100.0 | 186 | 1 | 11680 |
| \- globalstep[1] | 2.0 | 100.0 | 9 | 1 | 378 |
| \- ABM: Vacuum tubes | 12.6 | 4.8 | 114 | 72 | 215 |

rubenwardy commented:
How do you get .6 of a function call?

t4im commented:
Those are mean calls per server step in which the instrumented function was called at least once. (i.e. wasn't idle)

When saving to csv, you get the total amount of calls, but during `/profiler print` they are provided as average.

Wit ABM's for example it should equal the average amount of nodes being modified at once by it.

paramat commented:
Author disappeared, needs adoption.

paramat commented:
Close in 1 month if no adoption commitment.

paramat commented:
Neglected, closing.

paramat commented:
Seems to be waiting for reviews.

paramat commented:
Author is very absent, needs adoption and rebase.

paramat commented:
No adoption interest shown, closing for now.

t4im commented:
I offer to rebase it these days, if there are two devs offering to review it some time after (and gladly anyone else that is willing to).

You might want the first commit, if you plan to extend the profiler to measure more things in the future (e.g. i originally planned for experimental memory measurements against instrumentalized functions to enable people to find leaking mods, essentially addressing #2988 from the mod-side).

t4im commented:
Rebased with additional code-commentary and small fixes to the json/lua profile export.

SmallJoker commented on builtin/profiler/data.lua:
```diff
@@ -0,0 +1,143 @@
+-- a collection of statistics gained through instrument sampling
+local Stats = {
+	get_mean = function(self, key)
+		return self[key] and self[key]/self.samples
```
Space around binary operator

SmallJoker commented on builtin/profiler/data.lua:
```diff
@@ -0,0 +1,143 @@
+	end,
+
+	get_use = function(self)
+		return self.samples/ins_total.samples
```
Space around binary operator

SmallJoker commented on builtin/profiler/data.lua:
```diff
@@ -0,0 +1,143 @@
+local type, tostring = type, tostring
+local huge, pairs = math.huge, pairs
+
+local ins_total
```
Comment would be helpful. How about this one?
`-- Localized table ref to the currently running profiler`

SmallJoker commented on builtin/profiler/instrumentation.lua:
```diff
@@ -45,66 +47,128 @@ local register_functions = {
+-- An instrumented component may either be an instrumented function or a collection of instruments.
+-- Effectively creating a flat hierachy of instruments (usually with the top representing the entirety of instruments for a mod)
+-- References are passed around the profiler to avoid table lookups to find the instruments for logging,
+-- moving search and sort from instrumentation to sampling time for better performance.
```
These comments are quite long. Could you please limit each line to 80 (maximal 90) characters counting from the first non-whitespace?

SmallJoker commented on builtin/profiler/instrumentation.lua:
```diff
@@ -45,66 +47,128 @@ local register_functions = {
+local counts = {}
+local function generate_label(class, name, func_name)
+	local modname = get_current_modname()
+	if func_name then
```
Lua's conditional expression would probably be easier to read:
```Lua
func_name = func_name and func_name:gsub("^register_", "", 1) or ""
```

SmallJoker commented on builtin/profiler/reporter.lua:
```diff
@@ -73,105 +84,131 @@ local Formatter = {
+	update = function(self)
+		local widths = self.widths
+		if not widths then
+			widths = { 30, 5, 5, 9, 9, 9 }
```
1) Needs clarifying what these magic values do - and what they're used for (`-- Specifies each column width of the table output in characters`)
2) Shorter:
```Lua
self.widths = self.widths or { 30, 5, 5, 9, 9, 9 }
local widths = self.widths
```
Is localizing the variable needed here? I don't think that it's called so often that these microseconds actually bring a benefit.

t4im replied:
Not performance critical code, so not needed; just a habit. :-)

I'd still keep it localized here for the loop over `ins_stats` (grows with active instruments)
and having seen someone run the txt dump every few serverstep; it probably won't hurt.

SmallJoker commented:
The new code works nicely, but I'd also like to see an improvement in the txt format. Like this:
```
calls: Mean calls per server step in which the instrumented function was not idle
use %: Some kind of value which is close to either 0 or 100
avg µs: Average time spent per call
min µs: Minimal time spent per call
max µs: Maximal time spent per call
```

t4im commented:
Rebased against current master.
Also added a trivial protective fallback name for `get_current_modname` to `*unknown*` based on a report in minetest-hub by luk3yx of that happening in some on-the-fly definitions.

Fixer-007 commented:
This PR is almost 2 years old sitting with one approval, anyone with courage to sieve through PR's like that? Why is it on hold?

paramat commented:
Fixer that should be obvious, no need to ask, same reason as all the other delays.

Personally speaking, unfortunately it's not high enough priority for me to attend to until after 5.0.0 release. It's also a big PR and an area of code i'm not much good at yet.

t4im commented:
You are likely referring to `Data.reset()` (contrary to `Data:export` etc), which is not style related. If you look closer, `reset` isn't able to work on `self`, but works on shared states. Perhaps consider it a static method.

As for the preferred Lua-OOP style; that changed over the long course of the profiler and its PRs (passed from generation to generation of Tims) and might indeed have done better with a bit more consistency.
In the end i didn't want to risk larger changes for something like it though and are also not convinced you would get it completely consistent without some trade-offs, considering the supporting use of different paradigms in the code.

nerzhul commented:
oh it's a lua only profiler refactor. Nice. Code sounds good but i'm not the best in Lua. Pushing the PR to the 5.1.0 roadmap

rubenwardy commented:
rebased: https://github.com/rubenwardy/minetest/tree/profiler/data

SmallJoker commented:
@rubenwardy Feel free to force push it to this branch. Developers have write access to published PR branches.

rubenwardy commented:
I wasn't allowed to push,I think they disabled that

SmallJoker commented:
Indeed.
```
me@x7:/media/ramdisk/minetest$ LANG=en git push --force origin HEAD:profiler/data 
Username for 'https://github.com': SmallJoker
Password for 'https://SmallJoker@github.com': 
Counting objects: 13, done.
Delta compression using up to 4 threads.
Compressing objects: 100% (3/3), done.
Writing objects: 100% (13/13), 4.06 KiB | 4.06 MiB/s, done.
Total 13 (delta 10), reused 13 (delta 10)
remote: Resolving deltas: 100% (10/10), completed with 7 local objects.
To https://github.com/t4im/minetest.git
 ! [remote rejected]   HEAD -> profiler/data (permission denied)
error: failed to push some refs to 'https://github.com/t4im/minetest.git'
``^

paramat commented:
@t4im

sfan5 commented:
@t4im rebase still required
